motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,485 @@
+#include "motion_primitives_forward_controller/execution_state.hpp"
+#include "motion_primitives_forward_controller/motion_type.hpp"
+#include "motion_primitives_forward_controller/ready_for_new_primitive.hpp"
+
+namespace
+{  // utility
+
+// TODO(destogl): remove this when merged upstream
+// Changed services history QoS to keep all so we don't lose any client service calls
+static constexpr rmw_qos_profile_t rmw_qos_profile_services_hist_keep_all = {
+  RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+  1,  // message queue depth
+  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
+  RMW_QOS_POLICY_DURABILITY_VOLATILE,
+  RMW_QOS_DEADLINE_DEFAULT,
+  RMW_QOS_LIFESPAN_DEFAULT,
+  RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
+  RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
+  false};
+}  // namespace
+
+namespace motion_primitives_forward_controller
+{
+MotionPrimitivesForwardController::MotionPrimitivesForwardController()
+: controller_interface::ControllerInterface()
+{
+}
+
+controller_interface::CallbackReturn MotionPrimitivesForwardController::on_init()
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Initializing Motion Primitives Forward Controller");
+
+  try
+  {
+    param_listener_ =
+      std::make_shared<motion_primitives_forward_controller::ParamListener>(get_node());
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during controller's init with message: %s \n", e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn MotionPrimitivesForwardController::on_configure(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Configuring Motion Primitives Forward Controller");
+
+  params_ = param_listener_->get_params();
+
+  // Check if the name is not empty
+  if (params_.name.empty())
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "Error: A name must be provided!");
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // Check if there are exactly 25 command interfaces
+  if (params_.command_interfaces.size() != 25)
+  {  // motion_type + 6 joints + 2*7 positions + blend_radius + velocity + acceleration + move_time
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "Error: Exactly 25 command interfaces must be provided!");
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // Check if there are exactly 2 state interfaces
+  if (params_.state_interfaces.size() != 2)
+  {  // execution_status + ready_for_new_primitive
+    RCLCPP_ERROR(get_node()->get_logger(), "Error: Exactly two state interfaces must be provided!");
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // topics QoS
+  auto subscribers_qos = rclcpp::SystemDefaultsQoS();
+  subscribers_qos.keep_last(1);
+  subscribers_qos.best_effort();
+
+  // Reference Subscriber
+  ref_subscriber_ = get_node()->create_subscription<ControllerReferenceMsg>(
```

Comment:
I honestly think that the interface should be an action rather than a topic. (or, provide both, as the JTC does). That would also remove the need of the helpers you introduced in the interfaces package.

Furthermore, I think it makes sense to always provide a sequence as an interface. The sequence can, obviously, also have size 1. 

We have prepared the [`ExecuteMotion` action](https://github.com/UniversalRobots/industrial_robot_motion_interfaces/blob/main/action/ExecuteMotion.action) for that already.